source/s3_part_streaming_input_stream.c

```diff
@@ -0,0 +1,331 @@
+            AWS_ASSERT(impl->total_length_read <= impl->total_length);
+        }
+        aws_byte_buf_reset(impl->reading_chunk_buf, false);
+        impl->in_chunk_offset = 0;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_part_streaming_input_stream_get_status(struct aws_input_stream *stream, struct aws_stream_status *status) {
+    (void)stream;
+    (void)status;
+
+    struct aws_s3_part_streaming_input_stream_impl *impl =
+        AWS_CONTAINER_OF(stream, struct aws_s3_part_streaming_input_stream_impl, base);
+
+    status->is_end_of_stream = (impl->total_length_read == impl->total_length) || impl->eos_reached;
+    status->is_valid = true;
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_part_streaming_input_stream_get_length(struct aws_input_stream *stream, int64_t *out_length) {
+    AWS_ASSERT(stream != NULL);
+    struct aws_s3_part_streaming_input_stream_impl *impl =
+        AWS_CONTAINER_OF(stream, struct aws_s3_part_streaming_input_stream_impl, base);
+    *out_length = (int64_t)impl->total_length;
+    return AWS_OP_SUCCESS;
+}
+
+static void s_part_streaming_input_stream_destroy(void *user_data) {
+    struct aws_s3_part_streaming_input_stream_impl *impl = user_data;
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_S3_GENERAL,
+        "id=%p: Destroying part streaming input stream - total_read=%zu, total_length=%zu",
+        (void *)impl,
+        impl->total_length_read,
+        impl->total_length);
+
+    if (impl->loading_future) {
+        /* If there is a loading future, wait for it to complete. */
+        /* TODO: probably better to cancel the future, but we don't support cancel yet */
+        AWS_LOGF_DEBUG(AWS_LS_S3_GENERAL, "id=%p: Waiting for pending load to complete before destroy", (void *)impl);
+        aws_future_bool_wait(impl->loading_future, MAX_TIMEOUT_NS_P);
+        aws_future_bool_release(impl->loading_future);
+    }
+
+    aws_parallel_input_stream_release(impl->para_stream);
+    aws_s3_buffer_ticket_release(impl->ticket);
+    aws_mem_release(impl->allocator, impl);
+}
+
+static struct aws_input_stream_vtable s_part_streaming_input_stream_vtable = {
+    .seek = s_part_streaming_input_stream_seek,
+    .read = s_part_streaming_input_stream_read,
+    .get_status = s_part_streaming_input_stream_get_status,
+    .get_length = s_part_streaming_input_stream_get_length,
+};
+
+struct aws_input_stream *aws_part_streaming_input_stream_new(
+    struct aws_allocator *allocator,
+    struct aws_parallel_input_stream *para_stream,
+    struct aws_s3_buffer_ticket *buffer_ticket,
+    uint64_t offset,
+    size_t request_body_size) {
+    AWS_PRECONDITION(para_stream);
+    AWS_PRECONDITION(buffer_ticket);
+
+    struct aws_s3_part_streaming_input_stream_impl *impl =
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_s3_part_streaming_input_stream_impl));
+
+    aws_ref_count_init(
+        &impl->base.ref_count, impl, (aws_simple_completion_callback *)s_part_streaming_input_stream_destroy);
```

Comment:
thats probably not the most UB safe way (we had a bunch of prs to fix in other places). 
s_part_streaming_input_stream_destroy should probably just takes voids and cast